agent/consul/stream/event_publisher.go

```diff
@@ -157,56 +153,55 @@ func (e *EventPublisher) getTopicBuffer(topic Topic) *eventBuffer {
 // When the caller is finished with the subscription for any reason, it must
 // call Subscription.Unsubscribe to free ACL tracking resources.
 func (e *EventPublisher) Subscribe(req *SubscribeRequest) (*Subscription, error) {
-	// Ensure we know how to make a snapshot for this topic
-	_, ok := e.snapshotHandlers[req.Topic]
+	handler, ok := e.snapshotHandlers[req.Topic]
 	if !ok || req.Topic == nil {
 		return nil, fmt.Errorf("unknown topic %v", req.Topic)
 	}
 
 	e.lock.Lock()
 	defer e.lock.Unlock()
 
-	// Ensure there is a topic buffer for that topic so we start capturing any
-	// future published events.
-	buf := e.getTopicBuffer(req.Topic)
+	topicHead := e.getTopicBuffer(req.Topic).Head()
 
-	// See if we need a snapshot
-	topicHead := buf.Head()
-	var sub *Subscription
-	if req.Index > 0 && len(topicHead.Events) > 0 && topicHead.Events[0].Index == req.Index {
-		// No need for a snapshot, send the "end of empty snapshot" message to signal to
-		// client its cache is still good, then follow along from here in the topic.
+	// If the client view is fresh, resume the stream.
+	if req.Index > 0 && topicHead.HasEventIndex(req.Index) {
 		buf := newEventBuffer()
+		subscriptionHead := buf.Head()
+		// splice the rest of the topic buffer onto the subscription buffer so
+		// the subscription will receive new events.
+		buf.AppendItem(topicHead.NextLink())
+		return e.subscriptions.add(req, subscriptionHead), nil
+	}
 
-		// Store the head of that buffer before we append to it to give as the
-		// starting point for the subscription.
-		subHead := buf.Head()
+	snapFromCache := e.getCachedSnapshotLocked(req)
+	if req.Index == 0 && snapFromCache != nil {
+		return e.subscriptions.add(req, snapFromCache.First), nil
+	}
+	snap := newEventSnapshot()
 
-		buf.Append([]Event{{
-			Index:   req.Index,
+	// if the request has an Index the client view is stale and must be reset
+	// with a NewSnapshotToFollow event.
+	if req.Index > 0 {
+		snap.buffer.Append([]Event{{
 			Topic:   req.Topic,
 			Key:     req.Key,
-			Payload: endOfEmptySnapshot{},
+			Payload: newSnapshotToFollow{},
 		}})
 
-		// Now splice the rest of the topic buffer on so the subscription will
-		// continue to see future updates in the topic buffer.
-		buf.AppendItem(topicHead.NextLink())
-
-		sub = newSubscription(req, subHead, e.subscriptions.unsubscribe(req))
-	} else {
-		snap, err := e.getSnapshotLocked(req, topicHead)
-		if err != nil {
-			return nil, err
+		if snapFromCache != nil {
+			snap.buffer.AppendItem(snapFromCache.First)
+			return e.subscriptions.add(req, snap.First), nil
 		}
-		sub = newSubscription(req, snap.Head, e.subscriptions.unsubscribe(req))
 	}
 
-	e.subscriptions.add(req, sub)
-	return sub, nil
+	snap.appendAndSplice(*req, handler, topicHead)
+	e.setCachedSnapshotLocked(req, snap)
+	return e.subscriptions.add(req, snap.First), nil
```

Comment:
🤔 Hmmm. I might be wrong but I _think_ the refactor while the behaviour seems correct is likely negating some or all of the performance benefit of snapshot caching.

Caching snapshots is only really beneficial for thundering herds of clients requesting the same new snapshot at the same time. In the old code we measured this to be really significant when the snapshot is large because we avoid doing the work of querying and building the snapshot for every client since we generate the snapshot within `getCachedSnapshotLocked` while holding the lock. In other words only one request of the herd does the work while others wait and then once the snapshot is created the same snapshot is used for all of the waiting subscriptions.

As far as I can see this logic is not quite the same - now we might hit this code block concurrently from thousands of goroutines and each one will end up doing the snapshot work redundantly.

The cache still works for subsequent request accepted after at least one of them completes so it's not entirely negated but since it's main benefit is during highly concurrent accesses I think this could be significant. It's not worth drastically complicating the code for unless we benchmark and prove it to be a significant issue but if it's possible to not loose the "single flight" property of generating a snapshot for multiple concurrent callers that would be good.

I commented below that now this is syncronous here it's even worse in that a slowish snapshot with thousands of events might take say a few milliseconds to generate (memDB can be surprisingly slow when it has to do thousands of joins thanks to the index reflection creating lots of allocations even for a read - this happens in large service health results!) So the window between the leading request completing its snapshot and writing the cache under lock might be in the order of milliseconds in which case tens or hundreds of concurrent requests might have also been doing the same redundant work.

Would it be possible instead to change how `getSnapshotLocked` works so that it still has "read through" behaviour so concurrent callers will block each other until the snapshot is made and then the rest of the code follows as here? I _think_ that only requires moving the `appendAndSplice` call back into `getCachedSnapshotLocked`. Doing more work inside the lock is usually a bad move for perf but here it should help avoid duplicate work from concurrent callers that would just be waiting for the same work to be done somewhere else with the previous design.

Reply:
I don't think the locking behaviour has changed, it should still be a single flight for all `Subscribe` calls.

In a previous PR I added a second lock, so there is now:
1. `EventPublisher.lock` which handlers synchronization of topic buffers, and the `snapCache` (document on the field)
2. `subscriptions.lock` which handles synchronization of `subscriptions.byToken`.

The locks must be acquired in that order to prevent deadlocks.

In `Subscribe` the `EventPublisher.lock` is taken on line `161` before doing any work. I believe that will ensure that only one snapshot is generated because other subscriptions will block here.

The `subscription.lock` that is used in `subscriptions.add` is only to synchronize `subscribe` with `unsubscribe` and `close`. In the unsubscirbe and close flows the `EventPublisher.lock` is not held.

Reply:
Ah cool. Yeah that sounds right. I missed the other lock above.